var/spack/repos/builtin/packages/kokkos-kernels/package.py

```diff
@@ -6,48 +6,261 @@
-    depends_on('kokkos@develop', when='@develop')
+    variant('debug', default=False, description="Build debug version of Kokkos")
 
-    patch('makefile.patch')
+    variant('serial', default=True, description="enable Serial backend (default)")
+    variant('pthreads', default=False, description="enable Pthreads backend")
+    variant('qthreads', default=False, description="enable Qthreads backend")
+    variant('cuda', default=False, description="enable Cuda backend")
+    variant('openmp', default=False, description="enable OpenMP backend")
 
-    def edit(self, spec, prefix):
-        makefile = FileFilter("src/Makefile")
-        makefile.filter('CXX = .*', 'CXX = ' + env['CXX'])
+    # Compilation options
+    variant('pic', default=False,
+            description="enable position independent code (-fPIC flag)")
 
-    def build(self, spec, prefix):
-        with working_dir('build', create=True):
-            makefile_path = '%s%s' % (self.stage.source_path, '/src/Makefile')
-            copy(makefile_path, 'Makefile')
-            make_args = [
-                'KOKKOSKERNELS_INSTALL_PATH=%s' % prefix,
-                'KOKKOSKERNELS_PATH=%s' % self.stage.source_path,
-                'KOKKOS_PATH=%s' % spec['kokkos'].prefix
-            ]
+    # Kokkos options
+    variant('aggressive_vectorization', default=False,
+            description="set aggressive_vectorization Kokkos option")
+    variant('disable_profiling', default=False,
+            description="set disable_profiling Kokkos option")
+    variant('disable_dualview_modify_check', default=False,
+            description="set disable_dualview_modify_check Kokkos option")
+    variant('enable_profile_load_print', default=False,
+            description="set enable_profile_load_print Kokkos option")
+    variant('compiler_warnings', default=False,
+            description="set compiler_warnings Kokkos option")
+    variant('disable_deprecated_code', default=False,
+            description="set disable_deprecated_code Kokkos option")
+    variant('enable_eti', default=False,
+            description="set enable_eti Kokkos option")
+
+    # CUDA options
+    variant('force_uvm', default=False,
+            description="set force_uvm Kokkos CUDA option")
+    variant('use_ldg', default=False,
+            description="set use_ldg Kokkos CUDA option")
+    variant('rdc', default=False,
+            description="set rdc Kokkos CUDA option")
+    variant('enable_lambda', default=False,
+            description="set enable_lambda Kokkos CUDA option")
+
+    host_values = ('AMDAVX', 'ARMv80', 'ARMv81', 'ARMv8-ThunderX',
+                   'Power7', 'Power8', 'Power9',
+                   'WSM', 'SNB', 'HSW', 'BDW', 'SKX',
+                   'KNC', 'KNL')
+
+    gpu_values = ('Kepler30', 'Kepler32', 'Kepler35', 'Kepler37',
+                  'Maxwell50', 'Maxwell52', 'Maxwell53',
+                  'Pascal60', 'Pascal61',
+                  'Volta70', 'Volta72')
```

Comment:
A lot of this duplication could be avoided by extending the `CudaPackage` base class